message_passing/waiting_for_handlers_and_compensation/workflows.py

```diff
@@ -0,0 +1,180 @@
+            return result
+        # 👉 Catch BaseException since asyncio.CancelledError does not inherit
+        # from Exception.
+        except BaseException as e:
+            if is_workflow_exit_exception(e):
+                self.workflow_exit.set_exception(e)
+                await workflow.wait_condition(workflow.all_handlers_finished)
+                await self.workflow_compensation()
+                self._workflow_compensation_done = True
+            raise
+
+    async def workflow_compensation(self):
+        await workflow.execute_activity(
+            activity_executed_to_perform_workflow_compensation,
+            start_to_close_timeout=timedelta(seconds=10),
+        )
+        self._update_compensation_done = True
+
+    @workflow.update
+    async def my_update(self) -> str:
+        """
+        An update handler that handles exceptions raised in its own execution
+        and in that of the main workflow method.
+
+        It ensures that:
+        - Compensation/cleanup is always performed when appropriate
+        - The update caller gets the update result, or WorkflowUpdateFailedError
+        """
+        # 👉 As with the main workflow method, the update application logic is
+        # implemented in a separate method in order to separate "platform-level"
+        # error-handling and compensation concerns from application logic. Note
+        # that coroutines must be wrapped in tasks in order to use
+        # workflow.wait.
+        update_task = asyncio.create_task(self._my_update())
+
+        # 👉 "Race" the workflow_exit future against the handler's own application
+        # logic. Always use `workflow.wait` instead of `asyncio.wait` in
+        # Workflow code: asyncio's version is non-deterministic.
+        await workflow.wait(  # type: ignore
+            [update_task, self.workflow_exit], return_when=asyncio.FIRST_EXCEPTION
+        )
+        try:
+            if update_task.done():
+                # 👉 The update has finished (whether successfully or not).
+                # Regardless of whether the main workflow method is about to
+                # exit or not, the update caller should receive a response
+                # informing them of the outcome of the update. So return the
+                # result, or raise the exception that caused the update handler
+                # to exit.
+                return await update_task
+            else:
+                # 👉 The main workflow method exited prematurely due to an
+                # error, and this happened before the update finished. Fail the
+                # update with the workflow exception as cause.
+                raise exceptions.ApplicationError(
+                    "The update failed because the workflow run exited"
+                ) from cast(BaseException, self.workflow_exit.exception())
+        # 👉 Catch BaseException since asyncio.CancelledError does not inherit
+        # from Exception.
+        except BaseException as e:
```

Comment:
I haven't thought deeply, but I'm wondering if using a child workflow would work better for use cases in which there are compensations.  Would it be more straightforward?  Would it work better, e.g. by letting the workflow exit and letting the child workflow finish up the compensations?
Obviously that's extra latency which some might not be able to afford.  And it's a lot of code, though it might be easier than this.